### PR TITLE
switch to StringTable for string caching

### PR DESCRIPTION
**Pull request by WalterBright**
Remove the wretched and incompetent StringTab hash table and replace it with StringTable. Being a hash table that actually works, far more redundant strings are combined. Also remove irrelevant assertexp_\* globals.

### CODE REVIEW

**Comment by yebblies** [comment]
Is there some way you can put in on the stack and clear it instead of new/delete?  This way will be annoying for ddmd, because stringtable is defined in D so the ctor isn't directly callable from C++.

**Comment by WalterBright** [comment]
I don't understand your suggestion.

File: src/e2ir.c [hunk 1]
```diff
@@ -834,32 +834,18 @@ elem *setArray(elem *eptr, elem *edim, Type *tb, elem *evalue, IRState *irs, int
...
 {
     //printf("clearStringTab()\n");
-    memset(stringTab, 0, sizeof(stringTab));
-    stidx = 0;
-
-    assertexp_sfilename = NULL;
-    assertexp_name = NULL;
-    assertexp_mn = NULL;
+    delete stringTab;
+    stringTab = new StringTable;
```
Comment (andralex) [inline src/e2ir.c]: This is inefficient - can't you reuse the same string table instead of deallocating and reallocating?

**Comment by andralex** [comment]
will pull in the interest of moving fast

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
> I don't understand your suggestion.

Try compiling ddmd, you'll see.

**Comment by WalterBright** [comment]
> Try compiling ddmd, you'll see.

Please just tell me.

**Comment by yebblies** [comment]
> stringtable is defined in D so the ctor isn't directly callable from C++.

**Comment by yebblies** [comment]
D doesn't emit constructors that are callable from C++, which is why all of the class construction from the glue layer has been changed to go through factory functions.  Same with destructors/delete.  It's all right, I'll work out some other way to do this.

File: src/e2ir.c (hunk 1 above)
Comment (WalterBright) [inline src/e2ir.c]: It's debatable. StringTable uses large pools, and they'll only be reallocated between object files being generated. The design of StringTable would have to be altered somewhat to support resetting it.

**Comment by WalterBright** [comment]
So, is the `new StringTable()` a problem - there isn't a constructor for it. I can see about the destructor.

**Comment by WalterBright** [comment]
Feel free to make a PR for what you need, and I'll get it pulled.

**Comment by yebblies** [comment]
@WalterBright I've opened #4317 to fix ddmd.  You were correct that the constructor is not a problem, just the dtor.

**Pull request merged**